File: src/main/java/com/scottscmo/Event.java
package com.scottscmo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public final class Event {
    public static final String DATA_DIR = "dataDir";

    private static final Map<String, List<Consumer<String>>> handlers = new HashMap<>();
    private static final Map<String, String> lastValues = new HashMap<>();

    public static void emit(String eventName, String value) {
        lastValues.put(eventName, value);
        handlers.get(eventName).forEach(fn -> fn.accept(value));
    }

    public static void subscribe(String eventName, Boolean init, Consumer<String> handler) {
        List<Consumer<String>> listeners = handlers.getOrDefault(eventName, new ArrayList<>());
        listeners.add(handler);
        if (!handlers.containsKey(eventName)) {
            handlers.put(eventName, listeners);
        }

        if (init) {
            String value = lastValues.get(eventName);
            if (value != null) {
                handler.accept(value);
            }
        }
    }

    public static void subscribe(String key, Consumer<String> handler) {
        subscribe(key, false, handler);
    }
}

File: src/main/kotlin/com/scottscmo/Event.kt
package com.scottscmo

typealias EventHandler = (value: Any) -> Unit

object Event {
    const val DATA_DIR = "dataDir"

    private val handlers: MutableMap<String, MutableList<EventHandler>> = mutableMapOf()
    private val lastValues: MutableMap<String, Any> = mutableMapOf()

    fun emit(eventName: String, value: Any) {
        lastValues[eventName] = value
        handlers[eventName]?.forEach { it(value) }
    }

    fun subscribe(eventName: String, init: Boolean, handler: EventHandler) {
        val listeners = handlers.getOrDefault(eventName, mutableListOf())
        listeners.add(handler)
        handlers[eventName] = listeners

        if (init) {
            val value = lastValues[eventName]
            if (value != null) {
                handler(value)
            }
        }
    }

    fun subscribe(key: String, handler: EventHandler) {
        subscribe(key, false, handler)
    }
}
